todo_app/app.py

```diff
@@ -1,92 +1,223 @@
       return redirect('/')
 
 
    @app.route('/items/start', methods = ['POST'])
+   @login_required
    def start_item():
-      if request.method == 'POST':
-         form_data = request.form
-         task_id = form_data["id"]
-         mongo_client.start_task(task_id)
-         
+      if (disable_login or current_user.role == writer_role):
+         if request.method == 'POST':
+            form_data = request.form
+            task_id = form_data["id"]
+            mongo_client.start_task(task_id)
+      else:
+         print("Reader Role is not allowed to start a task")   
       return redirect('/')
 
+   @app.route('/users/', methods = ['GET'])
+   @login_required
+   def get_users():
+      if (disable_login or current_user.role == admin_role):
+         
+         users = mongo_client.get_all_users()
+         
+         return render_template('users.html', title='User Admin',
+            users=users)
+      else:
+         print(f"This page is only for admin users")
+
+   @app.route('/users/update', methods = ['POST'])
+   @login_required
+   def update_user():
+      if (disable_login or current_user.role == admin_role):
+         if request.method == 'POST':
+            form_data = request.form
+            user_id_to_update = form_data["id"]
+            new_role = form_data["new_role"]
+            mongo_client.update_user(user_id_to_update, new_role)
+      else:
+         print("Reader Role is not allowed to start a task")   
+      return redirect('/users')
+
    @app.route('/items/restart', methods = ['POST'])
+   @login_required
    def restart_item():
-      if request.method == 'POST':
-         form_data = request.form
-         task_id = form_data["id"]
-         mongo_client.reopen_task(task_id)
-         
+      if (disable_login or current_user.role == writer_role):
+         if request.method == 'POST':
+               form_data = request.form
+               task_id = form_data["id"]
+               mongo_client.reopen_task(task_id)
+      else:
+         print("Reader Role is not allowed to restart a task")     
       return redirect('/')
 
+   @app.route('/login/callback', methods = ['GET'])
+   def login():
+      code = request.args.get("code")
+      
+      if request.method == 'GET':
+         code = request.args.get("code")
+         # Prepare and send a request to get tokens
+         token_url, headers, body = client.prepare_token_request(
+            "https://github.com/login/oauth/access_token",
+            authorization_response=request.url,
+            redirect_url=request.base_url,
+            code=code  
+         )
+
+         headers['Accept'] = 'application/json'
+
+         token_response = requests.post(
+            token_url,
+            headers=headers,
+            data=body,
+            auth=(oauth_client_id, oauth_secret_id)
+         )
+         
+         
+         client.parse_request_body_response(json.dumps(token_response.json()))
```

Comment:
Minor, but `token_response.text` will give you the text representation of the response straight up in the same way converting it to Json and back to text with `json.dumps(token_response.json())` will